core/drivers/imx/dcp/dcp.c

```diff
@@ -0,0 +1,776 @@
+	}
+
+	if ((k1[0] & BIT(7)) == 0) {
+		dcp_left_shift_buffer(k1, k2, 16);
+	} else {
+		dcp_left_shift_buffer(k1, tmp, 16);
+		dcp_xor(tmp, const_rb, k2, 16);
+	}
+
+	ret = TEE_SUCCESS;
+out:
+	dcp_cipher_do_final(&data);
+
+	return ret;
+}
+
+TEE_Result dcp_store_key(uint32_t *key, unsigned int index)
+{
+	uint32_t val = 0;
+	unsigned int i = 0;
+	uint32_t key_store_except = 0;
+
+	if (!key)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (index > DCP_SRAM_KEY_NB_SUBWORD - 1) {
+		EMSG("Bad parameters, index must be < %u",
+		     DCP_SRAM_KEY_NB_SUBWORD);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	key_store_except = cpu_spin_lock_xsave(&key_store_spinlock);
+
+	dcp_clk_enable(true);
+
+	val = DCP_SRAM_KEY_INDEX(index);
+	io_write32(dcp_base + DCP_KEY, val);
+
+	/*
+	 * Key is stored as four uint32 values, starting with subword0
+	 * (least-significant word)
+	 */
+	for (i = 0; i < DCP_SRAM_KEY_NB_SUBWORD; i++) {
+		val = TEE_U32_TO_BIG_ENDIAN(key[i]);
+		io_write32(dcp_base + DCP_KEYDATA, val);
+	}
+
+	dcp_clk_enable(false);
+
+	cpu_spin_unlock_xrestore(&key_store_spinlock, key_store_except);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result dcp_cmac(struct dcp_cipher_init *init, uint8_t *input,
+		    size_t input_size, uint8_t *output)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	uint8_t key1[DCP_AES128_KEY_SIZE] = { };
+	uint8_t key2[DCP_AES128_KEY_SIZE] = { };
+	unsigned int nb_blocks = 0;
+	bool block_complete = false;
+	struct dcp_cipher_data data = { };
+	struct dcp_align_buf x = { };
+	struct dcp_align_buf y = { };
+	struct dcp_align_buf last = { };
+	unsigned int i = 0;
+	uint8_t offset = 0;
+
+	if (!input || !output || !init)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ret = dcp_cipher_do_init(&data, init);
+	if (ret != TEE_SUCCESS) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	ret = dcp_calloc_align_buf(&x, DCP_AES128_BLOCK_SIZE);
+	if (ret != TEE_SUCCESS) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	ret = dcp_calloc_align_buf(&y, DCP_AES128_BLOCK_SIZE);
+	if (ret != TEE_SUCCESS) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	ret = dcp_calloc_align_buf(&last, DCP_AES128_BLOCK_SIZE);
+	if (ret != TEE_SUCCESS) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	/* Generate CMAC subkeys */
+	ret = dcp_cmac_subkey_generation(init, key1, key2);
+	if (ret != TEE_SUCCESS)
+		goto out;
+
+	/* Get number of block */
+	nb_blocks = ROUNDUP(input_size, DCP_AES128_BLOCK_SIZE) /
+		    DCP_AES128_BLOCK_SIZE;
+	// nb_blocks = ROUNDUP(input_size, DCP_AES128_BLOCK_SIZE);
+	block_complete = nb_blocks && !(input_size % DCP_AES128_BLOCK_SIZE);
+	if (nb_blocks == 0)
+		nb_blocks = 1;
+
+	for (i = 0; i < nb_blocks - 1; i++) {
+		dcp_xor(x.data, input + offset, y.data, DCP_AES128_BLOCK_SIZE);
+		ret = dcp_cipher_do_update(&data, y.data, x.data,
+					   DCP_AES128_BLOCK_SIZE);
+		if (ret)
+			goto out;
+		offset += DCP_AES128_BLOCK_SIZE;
+	}
+
+	/* Process the last block */
+	memcpy(last.data, input + offset, DCP_AES128_BLOCK_SIZE);
```

Comment:
If `input_size` isn't a multiple of `DCP_AES128_BLOCK_SIZE` we're going to read beyond the end of `input` here.

Reply:
Fixed